sui_programmability/examples/defi/sources/pool.move

```diff
@@ -0,0 +1,309 @@
+    public(script) fun test_init_pool_(test: &mut Scenario) {
+        let (owner, _) = people();
+
+        next_tx(test, &owner); {
+            pool::init_for_testing(ctx(test));
+        };
+
+        next_tx(test, &owner); {
+            let pool_cap = test::take_owned<pool::PoolCreatorCap>(test);
+
+            pool::create_pool(
+                &pool_cap,
+                mint<BEEP>(1000000, ctx(test)),
+                mint<SUI>(1000000000, ctx(test)),
+                1000,
+                ctx(test)
+            );
+
+            test::return_owned(test, pool_cap);
+        };
+
+        next_tx(test, &owner); {
+            let lsp = test::take_owned<Coin<LSP<BEEP>>>(test);
+            let pool = test::take_shared<Pool<BEEP>>(test);
+            let pool_mut = test::borrow_mut(&mut pool);
+            let (amt_sui, amt_tok, lsp_supply) = pool::get_amounts(pool_mut);
+
+            assert!(Coin::value(&lsp) == 1000, 0);
+            assert!(lsp_supply == 1000, 0);
+            assert!(amt_sui == 1000000000, 0);
+            assert!(amt_tok == 1000000, 0);
+
+            test::return_owned(test, lsp);
+            test::return_shared(test, pool);
+        };
+    }
+
+    /// The other guy tries to exchange 5_000_000 sui for ~ 5000 BEEP,
+    /// minus the commission that is paid to the pool.
+    public(script) fun test_swap_sui_(test: &mut Scenario) {
+        test_init_pool_(test);
+
+        let (_, the_guy) = people();
+
+        next_tx(test, &the_guy); {
+            let pool = test::take_shared<Pool<BEEP>>(test);
+            let pool_mut = test::borrow_mut(&mut pool);
+
+            pool::swap_sui(pool_mut, mint<SUI>(5000000, ctx(test)), ctx(test));
+            test::return_shared(test, pool);
+        };
+
+        // Check the value of the coin received by the guy.
+        // Due to rounding problem the value is not precise
+        // (better on larger numbers).
+        next_tx(test, &the_guy); {
+            let beep = test::take_owned<Coin<BEEP>>(test);
+            assert!(Coin::value(&beep) > 4950, 1);
+            test::return_owned(test, beep);
+        };
+    }
+
+    /// The owner swaps back BEEP for SUI and expects an increase in price.
+    /// The sent amount of BEEP is 1000, initial price was 1 BEEP : 1000 SUI;
+    public(script) fun test_swap_tok_(test: &mut Scenario) {
+        test_swap_sui_(test);
+
+        let (owner, _) = people();
+
+        next_tx(test, &owner); {
+            let pool = test::take_shared<Pool<BEEP>>(test);
+            let pool_mut = test::borrow_mut(&mut pool);
+
+            pool::swap_token(pool_mut, mint<BEEP>(1000, ctx(test)), ctx(test));
+            test::return_shared(test, pool);
+        };
+
+        next_tx(test, &owner); {
+            let sui = test::take_owned<Coin<SUI>>(test);
+
+            // Actual win is 1005971, which is ~ 0.6% profit
+            assert!(Coin::value(&sui) > 1000000u64, 2);
+
+            test::return_owned(test, sui);
+        };
+    }
+
```

Comment:
One other good test might be to:
- Have two different folks put liquidity in the pool
- Have someone do a swap
- Have both of them take all their liquidity out
- Ensure that there are no coins left in the pool

Reply:
Done!